SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp

```diff
@@ -1363,6 +1365,138 @@ bool RunSchemaCompiler()
+		if (!Value->TryGet##Type(OutParam))                                                                                                \
+		{                                                                                                                                  \
+			UE_LOG(LogSpatialGDKSchemaGenerator, Error, TEXT("Failed to get %s as type %s"), TEXT(#Value), TEXT(#Type));                   \
+			return false;                                                                                                                  \
+		}                                                                                                                                  \
+	} while (false)
+
+#define SAFE_TRYGETFIELD(Value, Type, FieldName, OutParam)                                                                                 \
+	do                                                                                                                                     \
+	{                                                                                                                                      \
+		if (!Value->TryGet##Type##Field(TEXT(FieldName), OutParam))                                                                        \
+		{                                                                                                                                  \
+			UE_LOG(LogSpatialGDKSchemaGenerator, Error, TEXT("Failed to get field %s of type %s from %s"), TEXT(FieldName), TEXT(#Type),   \
+				   TEXT(#Value));                                                                                                          \
+			return false;                                                                                                                  \
+		}                                                                                                                                  \
+	} while (false)
+
+bool ExtractComponentSetFromSchemaJson(const FString& SchemaJsonPath, TMap<uint32, FComponentIDs>& OutComponentSetMap)
+{
+	TSharedPtr<FJsonValue> SchemaBundleJson;
+	{
+		TUniquePtr<FArchive> SchemaFile(IFileManager::Get().CreateFileReader(*SchemaJsonPath));
+		if (!SchemaFile)
+		{
+			return false;
+		}
+
+		TSharedRef<TJsonReader<char>> JsonReader = TJsonReader<char>::Create(SchemaFile.Get());
+
+		FJsonSerializer::Deserialize(*JsonReader, SchemaBundleJson);
+	}
+
+	const TSharedPtr<FJsonObject>* RootObject;
+
+	if (!SchemaBundleJson || !SchemaBundleJson->TryGetObject(RootObject))
+	{
+		UE_LOG(LogSpatialGDKSchemaGenerator, Error, TEXT("Invalid schema bundle file %s"), *SchemaJsonPath);
+		return false;
+	}
+
+	const TArray<TSharedPtr<FJsonValue>>* SchemaFiles;
+	SAFE_TRYGETFIELD((*RootObject), Array, "schemaFiles", SchemaFiles);
+
+	TMap<FString, uint32> ComponentMap;
+	TMap<uint32, TSet<FString>> ComponentRefSetMap;
```

Comment:
more of a design question really: could it make more sense to loop through the schema files twice, and build the component name -> ID map first, then on the second iteration build the out map directly, and avoid having to construct this map?

Reply:
I have not measured the time it takes to go through the file, but we are talking about a 5MB json file for the TestGym project.

Reply:
Sure, thought this wasn't meant to be about performance. The question was more could this be simpler- I understand why this map needs to be constructed if we only want to pass through once, but it adds complexity to do so and then produce the output map after. 

This isn't a strong opinion, so happy either way.